node/overseer/src/lib.rs

```diff
@@ -770,567 +756,892 @@ impl<CV, CB, CS, SD, AD, AR, BS, BD, P, PoVD, RA, AS, NB, CA, CG, CP, ApD, ApV,
-	) -> AllSubsystems<CV, CB, CS, SD, AD, AR, BS, BD, P, PoVD, RA, AS, NB, NEW, CG, CP, ApD, ApV, GS> {
-		AllSubsystems {
-			candidate_validation: self.candidate_validation,
-			candidate_backing: self.candidate_backing,
-			candidate_selection: self.candidate_selection,
-			statement_distribution: self.statement_distribution,
-			availability_distribution: self.availability_distribution,
-			availability_recovery: self.availability_recovery,
-			bitfield_signing: self.bitfield_signing,
-			bitfield_distribution: self.bitfield_distribution,
-			provisioner: self.provisioner,
-			pov_distribution: self.pov_distribution,
-			runtime_api: self.runtime_api,
-			availability_store: self.availability_store,
-			network_bridge: self.network_bridge,
-			chain_api,
-			collation_generation: self.collation_generation,
-			collator_protocol: self.collator_protocol,
-			approval_distribution: self.approval_distribution,
-			approval_voting: self.approval_voting,
-			gossip_support: self.gossip_support,
+	async fn recv(&mut self) -> SubsystemResult<FromOverseer<M>> {
+		loop {
+			// If we have a message pending an overseer signal, we only poll for signals
+			// in the meantime.
+			if let Some((needs_signals_received, msg)) = self.pending_incoming.take() {
+				if needs_signals_received <= self.signals_received.load() {
+					return Ok(FromOverseer::Communication { msg });
+				} else {
+					self.pending_incoming = Some((needs_signals_received, msg));
+
+					// wait for next signal.
+					let signal = self.signals.next().await
+						.ok_or(SubsystemError::Context(
+							"No more messages in rx queue to process"
+							.to_owned()
+						))?;
+
+					self.signals_received.inc();
+					return Ok(FromOverseer::Signal(signal))
+				}
+			}
+
+			let mut await_message = self.messages.next().fuse();
+			let mut await_signal = self.signals.next().fuse();
+			let signals_received = self.signals_received.load();
+			let pending_incoming = &mut self.pending_incoming;
+
+			// Otherwise, wait for the next signal or incoming message.
+			let from_overseer = futures::select! {
+				msg = await_message => {
+					let packet = msg
+						.ok_or(SubsystemError::Context(
+							"No more messages in rx queue to process"
+							.to_owned()
+						))?;
+
+					if packet.signals_received > signals_received {
+						// wait until we've received enough signals to return this message.
+						*pending_incoming = Some((packet.signals_received, packet.message));
+						continue;
+					} else {
+						// we know enough to return this message.
+						FromOverseer::Communication { msg: packet.message}
+					}
+				}
+				signal = await_signal => {
+					let signal = signal
+						.ok_or(SubsystemError::Context(
+							"No more messages in rx queue to process"
+							.to_owned()
+						))?;
+
+					FromOverseer::Signal(signal)
+				}
+			};
+
+			if let FromOverseer::Signal(_) = from_overseer {
+				self.signals_received.inc();
+			}
+
+			return Ok(from_overseer);
```

Comment:
tracking of the difference between received signal across subsystems adds a lot of complexity and not sure if it actually solves any problems

if a subsystem is busy processing messages and doesn't catch up with signals, it might be awaiting a message from another subsystem and sending it a bunch of signals wouldn't unblock it 

what we can do alternatively is use select_biased to prefer selecting signals to messages

Reply:
> tracking of the difference between received signal across subsystems adds a lot of complexity and not sure if it actually solves any problems

We don't do this, but yeah, I don't think that solves any problems.

I can switch to select_biased. But I believe that's not necessary as select will choose randomly if both streams are ready. The probability of selecting only messages is vanishingly small.

We use send_unbounded to break cycles so that awaiting on a subsystem won't stall. Although we will encode this in the type system in a future refactoring

Reply:
> We don't do this

Maybe I named it poorly, but that's what I meant we do with `pending_incoming`.

Reply:
We don't track any difference. What exactly is the issue you have with `pending_incoming`? I added that code to maintain the invariant that a message sent after a signal is received on subsystem A will only arrive after the same signal on subsystem B. This is an important property which our subsystems rely on heavily, so I won't remove it.

Reply:
> What exactly is the issue you have with pending_incoming?

Added complexity. I believe it can be achieved in a more simple way, e.g. with a 
https://docs.rs/futures/0.3.13/futures/lock/struct.Mutex.html
locked in `broadcast_signal` and `route_message`, can even be an `RwLock` with `write` in `broadcast_signal` and `read` in `route_message`. This will guarantee the ordering between signals and messages now that we use `select_biased!`.

One might argue that it will block overseer if the signal channel buffer is full for some subsystem, but in this impl it will indefinitely block the subsystem.

Reply:
although that doesn't work after we add direct communication, yeah, ok